Sources/MongoSwift/SDAM.swift

```diff
@@ -353,4 +375,64 @@ public struct TopologyDescription: Equatable {
         self.servers = size > 0 ? Array(buffer).map { ServerDescription($0!) } : []
         // the buffer is documented as always containing non-nil pointers (if non-empty).
     }
+
+    // For testing purposes
+    internal init(type: TopologyType, servers: [ServerDescription]) {
+        self.type = type
+        self.servers = servers
+    }
+}
+
+extension TopologyDescription {
+    internal func findSuitableServers(readPreference: ReadPreference? = nil) -> [ServerDescription] {
+        switch self.type._topologyType {
+        case .unknown:
+            return []
+        case .single,
+             .loadBalanced:
+            return self.servers
+        case .replicaSetNoPrimary,
+             .replicaSetWithPrimary:
+            switch readPreference?.mode {
+            case .secondary:
+                let secondaries = self.servers.filter { $0.type == .rsSecondary }
+                return self.replicaSetHelper(readPreference: readPreference, servers: secondaries)
+            case .nearest:
+                let secondariesAndPrimary = self.servers.filter { $0.type == .rsSecondary || $0.type == .rsPrimary }
+                return self.replicaSetHelper(readPreference: readPreference, servers: secondariesAndPrimary)
+            case .secondaryPreferred:
+                // If mode is 'secondaryPreferred', attempt the selection algorithm with mode 'secondary' and the
+                // user's maxStalenessSeconds and tag_sets.If no server matches, select the primary.
+                let secondaries = self.servers.filter { $0.type == .rsSecondary }
+                let primaries = self.servers.filter { $0.type == .rsPrimary }
+                let matches = self.replicaSetHelper(readPreference: readPreference, servers: secondaries)
+                return matches.isEmpty ? primaries : matches
+            case .primaryPreferred:
+                // If mode is 'primaryPreferred' or a readPreference is not provided, select the primary if it is known,
+                // otherwise attempt the selection algorithm with mode 'secondary' and the user's
+                // maxStalenessSeconds and tag_sets.
+                let primaries = self.servers.filter { $0.type == .rsPrimary }
+                if !primaries.isEmpty {
+                    return primaries
+                }
+                let secondaries = self.servers.filter { $0.type == .rsSecondary }
+                return self.replicaSetHelper(readPreference: readPreference, servers: secondaries)
+            default:
+                // the default mode is 'primary'.
+                return self.servers.filter { $0.type == .rsPrimary }
+            }
+        case .sharded:
+            return self.servers.filter { $0.type == .mongos }
+        }
+    }
+
+    internal func replicaSetHelper(
```

Comment:
I think it would be beneficial to rename this method to something more specific. That way, readers of code that invokes this method can infer what it does without having to check its implementation. e.g. something like "filterReplicaSetServers" or something like that.

Reply:
thats a great idea, thanks!